routing/cross_mwm_index_graph.cpp

```diff
@@ -22,76 +49,104 @@ bool GetTransitionSegment(OsrmFtSegMapping const & segMapping, TWrittenNodeId no
       continue;
 
     CHECK_NOT_EQUAL(seg.m_pointStart, seg.m_pointEnd, ());
     segment = Segment(numMwmId, seg.m_fid, min(seg.m_pointStart, seg.m_pointEnd), seg.IsForward());
     return true;
   }
-  LOG(LERROR, ("No valid segments in the range returned by OsrmFtSegMapping::GetSegmentsRange(", nodeId,
+  LOG(LDEBUG, ("No valid segments in the range returned by OsrmFtSegMapping::GetSegmentsRange(", nodeId,
                "). Num mwm id:", numMwmId));
   return false;
 }
 
-void AddTransitionSegment(OsrmFtSegMapping const & segMapping, TWrittenNodeId nodeId,
-                          NumMwmId numMwmId, std::vector<Segment> & segments)
+void AddFirstValidSegment(OsrmFtSegMapping const & segMapping, NumMwmId numMwmId,
+                          TWrittenNodeId nodeId, vector<Segment> & segments)
 {
   Segment key;
-  if (GetTransitionSegment(segMapping, nodeId, numMwmId, key))
+  if (GetFirstValidSegment(segMapping, numMwmId, nodeId, key))
     segments.push_back(key);
 }
 
 void FillTransitionSegments(OsrmFtSegMapping const & segMapping, TWrittenNodeId nodeId,
                             NumMwmId numMwmId, ms::LatLon const & latLon,
-                            std::map<Segment, ms::LatLon> & transitionSegments)
+                            std::map<Segment, std::vector<ms::LatLon>> & transitionSegments)
 {
   Segment key;
-  if (!GetTransitionSegment(segMapping, nodeId, numMwmId, key))
+  if (!GetFirstValidSegment(segMapping, numMwmId, nodeId, key))
     return;
 
-  auto const p = transitionSegments.emplace(key, latLon);
-  if (p.second)
-    return;
-
-  // @TODO(bykoianko) It's necessary to investigate this case.
-  LOG(LWARNING, ("Trying to insert a transition segment that was previously inserted. The key:",
-                 *p.first, ", the value:", latLon));
+  transitionSegments[key].push_back(latLon);
 }
 
-ms::LatLon const & GetLatLon(std::map<Segment, ms::LatLon> const & segMap, Segment const & s)
+vector<ms::LatLon> const & GetLatLon(std::map<Segment, std::vector<ms::LatLon>> const & segMap,
+                                     Segment const & s)
 {
   auto it = segMap.find(s);
   CHECK(it != segMap.cend(), ());
   return it->second;
 }
+
+void AddSegmentEdge(NumMwmIds const & numMwmIds, OsrmFtSegMapping const & segMapping,
+                    CrossWeightedEdge const & osrmEdge, bool isOutgoing, NumMwmId numMwmId,
+                    vector<SegmentEdge> & edges)
+{
+  BorderCross const & target = osrmEdge.GetTarget();
+  CrossNode const & crossNode = isOutgoing ? target.fromNode : target.toNode;
+
+  if (!crossNode.mwmId.IsAlive())
+    return;
+
+  NumMwmId const crossNodeMwmId = numMwmIds.GetId(crossNode.mwmId.GetInfo()->GetLocalFile().GetCountryFile());
+  CHECK_EQUAL(numMwmId, crossNodeMwmId, ());
+
+  Segment segment;
+  if (!GetFirstValidSegment(segMapping, crossNodeMwmId, crossNode.node, segment))
+    return;
+
+  // OSRM and AStar have different car models, therefore AStar heuristic doen't work for OSRM edges.
```

Comment:
// OSRM and AStar have different car models, therefore AStar heuristic doesn't work for OSRM node ids (edges).
  // This factor makes index gpaph (used in AStar) edge weight smaller then node ids weight.